Render VMs power state as icon with tooltip

rszwajko opened the pull request:
![Screenshot from 2023-09-08 19-02-34](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/03e488b6-69ba-4f74-bb36-92570dee047c)
![Screenshot from 2023-09-08 19-02-24](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/3fd81326-7561-43fa-b723-384494b1a286)

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/tabs/VirtualMachines/components/PowerStateCellRenderer.tsx:
```diff
@@ -0,0 +1,26 @@
+
+import { VMCellProps } from './VMCellProps';
+
+export const PowerStateCellRenderer: React.FC<VMCellProps> = ({ data }) => {
```
This renderer is specific to vmWarre, OCP status will may different ( it has more options that on/off )
If you choose to make a different status rendered for each type, this should have a unique name of path to indicate it's vmware onle.

For example if you choose unique name it can be VSpherePowerCellRendere, if you choose path it can be /cellRenderers/VSphere/powerCellRendere ?

rszwajko replied:
Switched to generic mapper borrowed from legacy package https://github.com/kubev2v/forklift-console-plugin/blob/8c17951019a08a33f55b7c820e694551dd99c6dc/packages/legacy/src/common/components/VMNameWithPowerState.tsx#L24

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/tabs/VirtualMachines/components/PowerStateCellRenderer.tsx:
```diff
@@ -0,0 +1,26 @@
+
+  return (
+    <TableCell>
+      <Tooltip content={tooltipText}>{icon}</Tooltip>
```
Icon only may not be clear, we can add text, for example "<On Icon> On" and "<Off Icon> Off", WDYT ?

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/tabs/VirtualMachines/components/PowerStateCellRenderer.tsx:
```diff
@@ -0,0 +1,26 @@
+export const PowerStateCellRenderer: React.FC<VMCellProps> = ({ data }) => {
+  const powerState = (data?.vm as VSphereVM)?.powerState;
+  const { t } = useForkliftTranslation();
+  const [icon, tooltipText] =
+    powerState === 'poweredOn'
+      ? [<PowerOffIcon color="green" key="on" />, t('Powered on')]
+      : powerState === 'poweredOff'
+      ? [<OffIcon color="red" key="off" />, t('Powered off')]
+      : [<UnknownIcon key="unknown" />, t('Unknown power state')];
```
```suggestion
const powerStateDetails = {
  poweredOn: {
    icon: <PowerOffIcon color="green" key="on" />,
    tooltip: t('Powered on')
  },
  poweredOff: {
    icon: <OffIcon color="red" key="off" />,
    tooltip: t('Powered off')
  },
  default: {
    icon: <UnknownIcon key="unknown" />,
    tooltip: t('Unknown power state')
  }
};

const { icon, tooltip } = powerStateDetails[powerState] || powerStateDetails.default;
```

`powerStateDetails` can be defined elsewhere, or you can even make a helper method `getPowerStateDetails(powerState: string) `

yaacov replied:
the `getPowerStateDetails` has an aditional advantage, it can make this method work for more then one type, as it can return other icons and texts for other types of providers depending on there entire object and not just the powerState field

rszwajko replied:
The code has been extracted to generic method (borrowed from [the legacy](https://github.com/kubev2v/forklift-console-plugin/blob/8c17951019a08a33f55b7c820e694551dd99c6dc/packages/legacy/src/common/components/VMNameWithPowerState.tsx#L24)).  Let me know if we need to split it more.

yaacov commented:
Ref: https://github.com/kubev2v/forklift-console-plugin/issues/719

rszwajko commented:
Newest force push add dependency on #728 

1. added displaying power state for all providers except OVA 
2. fixed filtering by status

![image](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/731dfc08-a463-460c-906d-737e8954d185)

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/tabs/VirtualMachines/utils/helpers/getVmPowerState.ts:
```diff
@@ -0,0 +1,58 @@
+export type PowerState = 'on' | 'off' | 'unknown';
+
+// moved from packages/legacy/src/common/components/VMNameWithPowerState.tsx
+export const getVmPowerState = (
```
```suggestion
export const getVmPowerState = (vm?: ProviderVirtualMachine): PowerState => {
  if (!vm) return 'unknown';
  
  const providerType = getProviderTypeFromVm(vm);
  
  switch (providerType) {
    case 'ovirt':
      return getOvirtVmPowerState(vm);
    case 'vsphere':
      return getVSphereVmPowerState(vm);
    case 'openstack':
      return getOpenstackVmPowerState(vm);
    case 'openshift':
      return getOpenshiftVmPowerState(vm);
    case 'ova':
      return getOvaVmPowerState();
    default:
      return 'unknown';
  }
};

const getOvirtVmPowerState = (vm: ProviderVirtualMachine): PowerState => {
  const status = (vm as OVirtVM).status;
  if (status === 'up') return 'on';
  if (status === 'down') return 'off';
  return 'unknown';
};

const getVSphereVmPowerState = (vm: ProviderVirtualMachine): PowerState => {
  const powerState = (vm as VSphereVM).powerState;
  if (powerState === 'poweredOn') return 'on';
  if (powerState === 'poweredOff') return 'off';
  return 'unknown';
};

const getOpenstackVmPowerState = (vm: ProviderVirtualMachine): PowerState => {
  const status = (vm as OpenstackVM).status;
  if (status === 'ACTIVE') return 'on';
  if (status === 'SHUTOFF') return 'off';
  return 'unknown';
};

const getOpenshiftVmPowerState = (vm: ProviderVirtualMachine): PowerState => {
  const status = (vm as OpenshiftVM)?.object?.status?.printableStatus;
  return status === 'Running' ? 'on' : 'off';
};

const getOvaVmPowerState = (): PowerState => {
  return 'off';
};

const getProviderTypeFromVm = (vm: ProviderVirtualMachine): ProviderType => {
  // Assuming this method exists and returns the correct type
  return getProviderType(vm);
};
```

My concerns about the  way this method is currently implemented:
a. if you  need the type, then it's not a universal method, we may not need it if we know the type when we call it.
b. some anonymous methods extract the status into a const befor using it, and some not
c. each anonymous method actually has a name, each get the power for a different provider, they can be extracted to individual methods with names that explain what they do.

yaacov replied:
p.s.

`getProviderTypeFromVm`  is not a real method, it's there to show how a universal method work, it should not need to get the type, if you need the type, you may not need this method at all, and you can just use the individual methods.

yaacov replied:
p.p.s

you can add a providerType to the vm interface, or to the https://github.com/kubev2v/forklift-console-plugin/blob/main/packages/types/src/types/provider/openstack/Resource.ts interface (if it't the parent of all provider types) and use it to know the provider type, see:
https://github.com/kubev2v/forklift-console-plugin/blob/main/packages/types/src/types/secret/OvaProviderSecret.ts#L9C3-L9C23

yaacov replied:
p.p.p.s
hmm, each provider will have different state nuances, best to return not just the state but alsoe the `icon`, `shortText` and `tooltip message`, so each provider can have it's own definition of what is on/off/migrating/sleeping etc. the text for openshift "on" maybe different from openstack "on" and the "off" for OVA will be different from vsphere "off"

rszwajko replied:
@yaacov 
>  if you need the type, then it's not a universal method, we may not need it if we know the type when we call it.
>

We need a way to chose between execution paths -  however if we extend the VM type to include `providerType` then we can branch on that.

> each provider will have different state nuances, best to return not just the state but alsoe the icon, shortText and tooltip message, so each provider can have it's own definition of what is on/off/migrating/sleeping etc. the text for openshift "on" maybe different from openstack "on" and the "off" for OVA will be different from vsphere "off"
>

We rely on common set of states (`PowerState` interface) for filtering. If we want to have different states then we need a power state enum per provider. If the differences are limited to the presentation layer (icon, tooltip,etc) then we can handle it in the renderer i.e.use different `states` dictionary. Currently a common mapping is used:
```
 const states: { [key in PowerState]: [JSX.Element, string, string] } = {
    on: [<PowerOffIcon color="green" key="on" />, t('Powered on'), t('On')],
    off: [<OffIcon color="red" key="off" />, t('Powered off'), t('Off')],
    unknown: [<UnknownIcon key="unknown" />, t('Unknown power state'), t('Unknown')],
  };
```

rszwajko replied:
Addressed in the newest version.

yaacov commented:
@rszwajko hi, can you update the screenshots and cleanup the commits ?

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/tabs/VirtualMachines/VSphereVirtualMachinesList.tsx:
```diff
@@ -67,7 +67,7 @@ export const vSphereVmFieldsMetadataFactory: ResourceFieldFactory = (t) => [
   {
     resourceFieldId: 'powerState',
-    jsonPath: (data) => getVmPowerState('vsphere', (data as VmData)?.vm),
+    jsonPath: (data) => getVmPowerState((data as VmData)?.vm),
```
jsonPath: (data: VmData) => getVmPowerState(data?.vm)
 
 looks nicer to me, can you try ?

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/tabs/VirtualMachines/utils/helpers/getVmPowerState.ts:
```diff
@@ -8,47 +8,57 @@ import {
-      powerStatus = 'unknown';
-    }
+    case 'ovirt':
+      return getOVirtVmPowerState(vm as OVirtVM);
```
we should not need the casting now, do we need to redo the typing PR ?

rszwajko commented:
Newest force push:
1. removed unnecessary casts (as pointed above)
2. added missing status column for OpenShift - got lost at some point

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/tabs/VirtualMachines/utils/helpers/index.ts:
```diff
@@ -1,3 +1,4 @@
 // @index(['./*', /style/g], f => `export * from '${f.path}';`)
 export * from './getHighestPriorityConcern';
+export * from './getVmPowerState';
```
nice :white_flower:
